src/util/codec/mysql/decimal.rs

```diff
@@ -0,0 +1,679 @@
+
+macro_rules! enable_conv_for_int {
+    ($t:ty) => {
+        impl From<$t> for Decimal {
+            fn from(t: $t) -> Decimal {
+                Decimal {
+                    coeff: t.into(),
+                    exp: 0,
+                    fsp: 0,
+                }
+            }
+        }
+    };
+}
+
+enable_conv_for_int!(u32);
+enable_conv_for_int!(u64);
+enable_conv_for_int!(u16);
+enable_conv_for_int!(u8);
+enable_conv_for_int!(i32);
+enable_conv_for_int!(i64);
+enable_conv_for_int!(i16);
+enable_conv_for_int!(i8);
+enable_conv_for_int!(usize);
+enable_conv_for_int!(isize);
+
+impl Decimal {
+    /// Convert a float number to decimal.
+    ///
+    /// This function will use float's canonical string representation
+    /// rather than the accurate value the float represent.
+    pub fn from_f64(f: f64) -> Result<Decimal> {
+        if !f.is_finite() {
+            return Err(invalid_type!("{} can't be convert to decimal'", f));
+        }
+
+        let s = format!("{}", f);
+        Decimal::from_str(&s)
+    }
+
+    /// Construct a decimal using the specified coefficient, exponent and fsp.
+    ///
+    /// If constucted decimal contains more fraction digit than fsp, they will be
+    /// rounded.
+    pub fn new(coeff: BigInt, exp: i32, mut fsp: usize) -> Decimal {
+        fsp = cmp::min(fsp, MAX_FSP);
+
+        let d = Decimal {
+            coeff: coeff,
+            exp: exp,
+            fsp: fsp,
+        };
+
+        if exp >= 0 {
+            return d;
+        }
+
+        if fsp as i32 + exp < 0 {
+            return d.rescale(-(fsp as i32));
+        }
+        d
+    }
+
+    /// Remove extra zero in `coeff`.
+    ///
+    /// For example, 1000 * 10<sup>3</sup> will be compact to 1 * 10<sup>6</sup>,
+    /// 1000 * 10<sup>-3</sup> will be compact to 1 * 10<sup>0</sup>.
+    pub fn compact(mut self) -> Decimal {
+        let ten = 10.into();
+        loop {
+            let res = self.coeff.div_rem(&ten);
+            if !res.1.is_zero() {
+                break;
+            }
+            self.coeff = res.0;
+            self.exp += 1;
+        }
+        self
+    }
+
+    /// Adjust decimal's exponent to specified value.
+    ///
+    /// This may cause the coefficient grows or decreases.
+    /// For example, 12345 * 10<sup>3</sup> rescale with 4 will be represented
+    /// as 1235 * 10<sup>4</sup>; 12345 * 10<sup>-3</sup> rescale with -5 will
+    /// be represented as 1234500 * 10<sup>-5</sup>
+    pub fn rescale(&self, mut exp: i32) -> Decimal {
+        if exp < 0 {
+            exp = cmp::max(exp, -(MAX_FSP as i32));
+        }
+        if self.exp == exp {
+            return self.clone();
+        }
+
+        let to_add = exp - self.exp;
+        let coeff = if to_add > 0 {
+            let divider = BigInt::from(num::pow(BigUint::from(10u8), to_add as usize));
+            let (mut coeff, rem) = self.coeff.div_rem(&divider);
+            if rem >= divider / BigInt::from(2) {
```

Comment:
Is this rounding? Please add some comment.
`rem * BigInt::from(2) >= divider` seems more clear and efficient.